torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -543,6 +587,9 @@ def _reset_lazy_init(self) -> None:
                 # reset attributes that are added in _init_param_attributes, as
                 # part of _lazy_init
                 del p._local_shard  # type: ignore[attr-defined]
+        # reset 'reshard_after_forward' flag as well, as it could be changed during
+        # training loop, e.g, it was set as False for root module in _lazy_init().
```

Comment:
should we support changing this during the training loop? 

Also, if it is set as `False` for root module during _lazy_init but this runs contradictory to the sharding strategy specified by the user, should we actually have that be respected by the root module in `_lazy_init`? Might be confusing to users otherwise.

Reply:
if users set reshard_after_forward = False, setting it as False for root module does not matter. if users set reshard_after_forward=True, internally setting it as True for root module is kind of internal optimization. Also reshard_after_forward is transparent to users, right? users do not need to know this optimization

Reply:
>  internally setting it as True for root module is kind of internal optimization. 

Sorry I might be a bit confused, but do you mean "internally setting it as False"? I believe we currently set reshard_after_forward = False for root module right? 

Maybe a comment explaining this might be useful?

Reply:
sorry internally setting it as False, refactored comments